src/main/java/bio/terra/stairway/Flight.java

```diff
@@ -154,9 +161,37 @@ private StepResult runSteps() throws InterruptedException, StairwayExecutionExce
                 return result;
             }
 
-            flightDao.step(context());
+            switch (result.getStepStatus()) {
+                case STEP_RESULT_SUCCESS:
+                    // Finished a step; run the next one
+                    context().setRerun(false);
+                    flightDao.step(context());
+                    context().nextStepIndex();
+                    break;
+
+                case STEP_RESULT_RERUN:
+                    // Rerun the same step
+                    context().setRerun(true);
+                    flightDao.step(context());
+                    break;
+
+                case STEP_RESULT_YIELD:
+                    // Finished a step; yield execution
+                    context().setRerun(false);
+                    flightDao.step(context());
+                    return result;
+
+                case STEP_RESULT_STOP:
+                    // Stop executing - leave rerun setting as is; we'll need to pick up where we left off
+                    flightDao.step(context());
+                    return result;
 
-            context().nextStepIndex();
+                case STEP_RESULT_FAILURE_RETRY:
+                case STEP_RESULT_FAILURE_FATAL:
+                default:
+                    // unreachable
```

Comment:
because there should always be an enum value?

Reply:
Because the `if (!result.isSuccess())` return above would have already returned.
I should throw if the Step ever returns a step result that is not in the Enum. I'll add that in.

Reply:
On that note, if either of the FAILURE cases get reached here should this throw an exception?

Reply:
I'm a fan of making even `unreachable` sections `throw`

Reply:
I changed it to throw.